diffkemp/compare.py

```diff
@@ -1,243 +1,301 @@
-        else:
-            group_dir = None
+    def __init__(self, args):
+        self.args = args
+        self.config = Config.from_args(args)
+        self.result = Result(Result.Kind.NONE, args.snapshot_dir_old,
+                             args.snapshot_dir_new, start_time=default_timer())
+        self.regex_pattern = re.compile(args.regex_filter) \
+            if args.regex_filter else None
+        self.output_dir = None
+
+        # Group variables, set later
+        self.group = None
+        self.group_name = None
+        self.group_dir = None
+        self.group_printed = False
+        self.result_graph = None
+        self.cache = None
+        self.module_cache = {}
+        self.modules_to_cache = None
+
+    def run(self):
+        # Set the output directory
+        self._set_output_dir()
+        return self._compare_snapshots()
+
+    def _set_output_dir(self):
+        if self.args.stdout:
+            return
 
-        result_graph = None
-        cache = SimpLLCache(mkdtemp())
-        module_cache = {}
+        if not self.args.output_dir:
+            self.output_dir = self._default_output_dir()
+            return
 
-        if args.enable_module_cache:
-            modules_to_cache = _get_modules_to_cache(group.functions.items(),
-                                                     group_name,
-                                                     config.snapshot_second,
-                                                     2)
+        self.output_dir = self.args.output_dir
+        if os.path.isdir(self.output_dir):
+            raise OutputDirExistsError(
+                "Error: output directory {} exists\n".format(self.output_dir))
+
+    def _default_output_dir(self):
+        """Name of the directory to put log files into."""
+        base_dirname = "diff-{}-{}".format(
+            os.path.basename(os.path.normpath(self.args.snapshot_dir_old)),
+            os.path.basename(os.path.normpath(self.args.snapshot_dir_new)))
+        if os.path.isdir(base_dirname):
+            suffix = 0
+            dirname = base_dirname
+            while os.path.isdir(dirname):
+                dirname = "{}-{}".format(base_dirname, suffix)
+                suffix += 1
+            return dirname
+        return base_dirname
+
+    def _compare_snapshots(self):
+        for group_name, group in sorted(self.config.snapshot_first
+                                        .fun_groups.items()):
+            self._set_group_variables(group, group_name)
+            self._compare_groups()
+
+        self._finalize_output()
+        return 0
+
+    def _set_group_variables(self, group, group_name):
+        self.group = group
+        self.group_name = group_name
+        self._set_group_dir()
+        self.group_printed = False
+        self._set_modules_to_cache_if_enabled()
+        self.result_graph = None
+        self.cache = SimpLLCache(mkdtemp())
+        self.module_cache = {}
+
+    def _set_group_dir(self):
```

Comment:
I think it would be much nicer if this returned the group dir instead of setting it directly (so it would be something like `_get_group_dir`) and above, we would do:
```
self.group_dir = self._get_group_dir()
```